sdk/search/Azure.Search.Documents/src/Options/SearchConditionalOptions.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Search.Documents
+{
+    /// <summary>
+    /// Options to customize Search service operations conditioned on <see cref="IfMatch"/> or <see cref="IfNoneMatch"/> ETags.
+    /// </summary>
+    public class SearchConditionalOptions : SearchRequestOptions
```

Comment:
After talking with @annelo-msft I may actually remove this type and use simple `bool` parameters similar to AppConfig. I need to think through the scenarios more. The same reasoning should apply here, though.

Reply:
Note that because the etag property is nullable - the deserializer for one reason or another may not deserialize it - we should probably throw an argument exception if `onlyIfUnchanged` is `true` but a model's etag property is null.

This also means a change for delete methods that currently take just a string, rather than requiring getting an entire model first before deleting it. To handle that case (and perhaps create-or-update operations), request options could take an optional etag property.

Reply:
I see a couple ways of handling this.

1. We put bool the boolean `OnlyIfUnchanged` property and optional etag in the new `SearchConditionalOptions`.
2. We only put the optional etag in the new `SearchConditionalOptions` and put the boolean `onlyIfUnchanged` parameter in the method with a default value (at least for .NET) of `false` so it's obvious.

Delete methods could take either a string name or resource, and any etag value specified in the request options would override any etag of the resource passed since they would have to have explicitly set it.

There doesn't seem much value to supporting `IfNoneMatch` explicitly, but should we ever need to we could add overloads. That was AppConfig's thought as well.

Reply:
See https://github.com/heaths/azure-sdk-for-net/pull/2 for an alternative design.

/cc @sima-zhu @xirzec @xiangyan99 @bryevdv @tg-msft @brjohnstmsft

Reply:
Wait, the etag is nullable? As opposed to not required?

Reply:
If the service didn't send it back, there's no content. Client deserialization treats it as nullable anyway (only deserializes if retunred), so I don't see a choice but to support it being nullable. If we don't make it nullable, then calling `ETag.ToString` ends up returning an invalid value.

Reply:
Under what circumstances would one (in reality) not get an ETag value in a response?

Reply:
I don't know - maybe not even valid (@brjohnstmsft?), but I'd rather be safe than sorry. This is similar to the #9686 discussion, though that is probably more likely in practice than this. But I don't see how making `ETag` nullable makes the experience worse. Some uses have to be nullable anyway (e.g. properties on `MatchConditions` class in Azure.Core), so making them all nullable actually makes passing them around easier.

Reply:
I took a quick look in the docs and didn't see an obvious answer.  `MatchConditions` has nullable properties because they're optional inputs.  I do wonder if there are valid service scenarios for optional outputs or it's just a bug in the swagger.

Reply:
I seem to recall that we return ETags conditionally, but I don't remember the details. I'll have to poke at the service and get back to you.

Reply:
`@odata.etag` is returned if you include `odata.metadata=minimal` or `odata.metadata=full` in the `Accept` header. Since the libraries should always set it to `minimal`, you can assume that the ETag is always there.

Reply:
We're not setting it to minimal currently. If this is something we should be doing, please open a new issue to track.

I don't understand, though, what this being nullable is hurting. For one, on *this class* they have to be nullable because they have to not be specified together, or maybe neither should be set. The *only* way you can do that is if they are nullable.

As for `ETag` properties on models, that's another matter. If we're always sure they are returned, fine. But if even 1 response deserializes without, sending the model back will result in a service failure since "<null>" is invalid. So, again, what is this hurting? It's resilient.

Reply:
Polymorphism will break if you don't set `odata.metadata` to `minimal`. I thought `SearchServiceClient` was already doing that, otherwise all tests would be failing. `SearchClient` (nee `SearchIndexClient`) can use `odata.metadata=none` since there's no polymorphic models or ETags there.

Reply:
Nowhere in code do we set it, and recorded test data shows no signs. Only the response comes back with,
```json
"Content-Type": "application/json; odata.metadata=minimal; odata.streaming=true",
```
Does the service default to minimal? I'll open a bug to track.

Reply:
I believe it does, but it's better to explicitly set it.